tests/capture/test_base_interpreter.py

```diff
@@ -0,0 +1,439 @@
+
+    def interpret_operation(self, op):
+        new_op = op.simplify()
+        if new_op is op:
+            new_op = new_op._unflatten(*op._flatten())
+            # if new op isn't queued, need to requeue op.
+        return new_op
+
+
+# pylint: disable=use-implicit-booleaness-not-comparison
+def test_env_and_initialized():
+    """Test that env is initialized at the start."""
+
+    interpreter = SimplifyInterpreter()
+    assert interpreter._env == {}
+    assert interpreter._op_math_cache == {}
+
+
+def test_primitive_registrations():
+    """Test that child primitive registrations dict's are not copied and do
+    not effect PlxprInterpreeter."""
+
+    class SimplifyInterpreterLocal(PlxprInterpreter):
+
+        def interpret_operation(self, op):
+            new_op = op.simplify()
+            if new_op is op:
+                # if new op isn't queued, need to requeue op.
+                new_op = new_op._unflatten(*op._flatten())
+            return new_op
+
+    assert (
+        SimplifyInterpreterLocal._primitive_registrations
+        is not PlxprInterpreter._primitive_registrations
+    )
+
+    @SimplifyInterpreterLocal.register_primitive(qml.X._primitive)
+    def _(self, *invals, **params):  # pylint: disable=unused-argument
+        print("in custom interpreter")
+        return qml.Z(*invals)
+
+    assert qml.X._primitive in SimplifyInterpreterLocal._primitive_registrations
+    assert qml.X._primitive not in PlxprInterpreter._primitive_registrations
+
+    @SimplifyInterpreterLocal()
+    def f():
+        qml.X(0)
+        qml.Y(5)
+
+    jaxpr = jax.make_jaxpr(f)()
+
+    with qml.queuing.AnnotatedQueue() as q:
+        jax.core.eval_jaxpr(jaxpr.jaxpr, [])
+
+    qml.assert_equal(q.queue[0], qml.Z(0))  # turned into a Y
+    qml.assert_equal(q.queue[1], qml.Y(5))  # mapped wire
+
+
+def test_overriding_measurements():
+    """Test usage of an interpreter with a custom way of handling measurements."""
+
+    class MeasurementsToSample(PlxprInterpreter):
+
+        def interpret_measurement_eqn(self, primitive, *invals, **params):
+            temp_mp = primitive.impl(*invals, **params)
+            return qml.sample(wires=temp_mp.wires)
+
+    @MeasurementsToSample()
+    @qml.qnode(qml.device("default.qubit", wires=2, shots=5))
+    def circuit():
+        return qml.expval(qml.Z(0)), qml.probs(wires=(0, 1))
+
+    res = circuit()
+    assert qml.math.allclose(res[0], jax.numpy.zeros(5))
+    assert qml.math.allclose(res[1], jax.numpy.zeros((5, 2)))
+
+    jaxpr = jax.make_jaxpr(circuit)()
+    assert (
+        jaxpr.eqns[0].params["qfunc_jaxpr"].eqns[0].primitive
+        == qml.measurements.SampleMP._wires_primitive
+    )
+    assert (
+        jaxpr.eqns[0].params["qfunc_jaxpr"].eqns[1].primitive
+        == qml.measurements.SampleMP._wires_primitive
+    )
+
+
+def test_setup_method():
+    """Test that the setup method can be used to initialized variables each call."""
+
+    class CollectOps(PlxprInterpreter):
```

Comment:
TapeInterpreter half-way done? :o Reading these tests is both fun and inspiring!

Reply:
#6343 just opened.